File: ios/Classes/DartProxyCipherAdapter.swift
import Flutter
import Foundation

/// Built-in adapter that proxies each read chunk to a pure-Dart
/// `DartCipherDelegate` over a dedicated FlutterBasicMessageChannel
/// (binary codec), named `secure_video_player/dart_cipher_<channelId>`.
///
/// Request frame: [1B direction: 0=decrypt, 1=encrypt][8B file offset, big-endian][payload].
/// Reply: transformed bytes (same length). On timeout / wrong length / error
/// the buffer is left untouched so the decoder surfaces a read error (matching
/// the built-in adapters' failure convention).
///
/// Threading: transform() runs on the resource-loader queue (playback) or the
/// crypto queue (file encrypt) — never the main thread. FlutterBasicMessageChannel
/// must be used from the main thread, so we dispatch the send to the main queue
/// and block the caller on a semaphore. Main stays free to deliver Dart's reply,
/// so there is no deadlock. A main-thread call fails fast instead of deadlocking.
final class DartProxyCipherAdapter: CipherAdapter {

    private let messenger: FlutterBinaryMessenger
    private var channel: FlutterBasicMessageChannel?
    private var timeout: TimeInterval = 5.0

    init(messenger: FlutterBinaryMessenger) {
        self.messenger = messenger
    }

    func initialize(params: [String: Any?]) throws {
        guard let channelId = params["channelId"] as? String else {
            throw CipherError.badParams("dartProxy requires a 'channelId' string")
        }
        if let ms = params["timeoutMs"] as? NSNumber { timeout = ms.doubleValue / 1000.0 }
        channel = FlutterBasicMessageChannel(
            name: "\(SvpProtocol.channelDartCipherPrefix)\(channelId)",
            binaryMessenger: messenger,
            codec: FlutterBinaryCodec.sharedInstance())
    }

    /// Playback (resource-loader) path: decrypt, non-throwing. On timeout /
    /// wrong length / error the buffer is left untouched so the decoder
    /// surfaces a read error (matching the built-in adapters' convention).
    func transform(_ buffer: inout Data, filePosition: Int64) {
        guard let reply = try? send(buffer, filePosition: filePosition, encrypt: false)
        else { return }
        buffer = reply
    }

    /// File-cryptor path: throwing so a proxy failure fails the whole job
    /// instead of writing raw (untransformed) bytes to the output.
    func transform(_ buffer: inout Data, filePosition: Int64, encrypt: Bool) throws {
        buffer = try send(buffer, filePosition: filePosition, encrypt: encrypt)
    }

    /// Sends [1B dir][8B offset BE][payload] and returns the same-length reply.
    /// Throws on main-thread misuse, missing channel, timeout, or short reply.
    private func send(_ buffer: Data, filePosition: Int64, encrypt: Bool) throws -> Data {
        let length = buffer.count
        if length == 0 { return buffer }
        guard !Thread.isMainThread else {
            assertionFailure("DartProxyCipherAdapter.transform must not run on the main thread")
            throw CipherError.transformFailed("dartProxy transform ran on the main thread")
        }
        guard let channel else {
            throw CipherError.transformFailed("dartProxy channel not initialized")
        }

        // Frame: [dir: 0=decrypt, 1=encrypt][offset 8B BE][payload].
        var request = Data(capacity: 9 + length)
        request.append(encrypt ? 1 : 0)
        var beOffset = filePosition.bigEndian
        withUnsafeBytes(of: &beOffset) { request.append(contentsOf: $0) }
        request.append(buffer)

        let semaphore = DispatchSemaphore(value: 0)
        var reply: Data?
        DispatchQueue.main.async {
            channel.sendMessage(request) { response in
                reply = response as? Data
                semaphore.signal()
            }
        }
        if semaphore.wait(timeout: .now() + timeout) == .timedOut {
            throw CipherError.transformFailed("dartProxy transform timed out")
        }
        guard let reply, reply.count == length else {
            throw CipherError.transformFailed(
                "dartProxy returned \(reply?.count ?? 0) bytes, expected \(length)")
        }
        return reply
    }
}

File: android/src/main/kotlin/com/hulkenstein/secure_video_player/DartProxyCipherAdapter.kt
package com.hulkenstein.secure_video_player

import android.os.Handler
import android.os.Looper
import io.flutter.plugin.common.BasicMessageChannel
import io.flutter.plugin.common.BinaryCodec
import io.flutter.plugin.common.BinaryMessenger
import java.io.IOException
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit

/**
 * Built-in adapter that proxies each read chunk to a pure-Dart
 * `DartCipherDelegate` over a dedicated BasicMessageChannel (BinaryCodec),
 * named `secure_video_player/dart_cipher_<channelId>`.
 *
 * Request frame: [1B direction: 0=decrypt, 1=encrypt][8B file offset, big-endian][payload].
 * Reply: transformed bytes (same length); null/short/timeout -> IOException.
 *
 * Threading: transform() is called on a background reader (ExoPlayer's load
 * thread, FileCryptor's executor, or a probe thread) — never the main thread.
 * BinaryMessenger.send must run on the platform (main) thread, so we post the
 * send to the main looper and block the caller thread on a latch. The main
 * thread stays free to deliver Dart's reply, so there is no deadlock. If
 * transform() is ever called ON the main thread we fail fast instead of
 * deadlocking against ourselves.
 */
class DartProxyCipherAdapter(
    private val messenger: BinaryMessenger,
) : CipherAdapter {

    private val mainHandler = Handler(Looper.getMainLooper())
    private lateinit var channel: BasicMessageChannel<ByteBuffer>
    private var timeoutMs = 5000L

    override fun init(params: Map<String, Any?>) {
        val channelId = params["channelId"] as? String
            ?: throw IllegalArgumentException("dartProxy requires a 'channelId' string")
        (params["timeoutMs"] as? Number)?.let { timeoutMs = it.toLong() }
        channel = BasicMessageChannel(
            messenger,
            "${SvpProtocol.CHANNEL_DART_CIPHER_PREFIX}$channelId",
            BinaryCodec.INSTANCE,
        )
    }

    override fun transform(buffer: ByteArray, offset: Int, length: Int, filePosition: Long) =
        transform(buffer, offset, length, filePosition, encrypt = false)

    override fun transform(
        buffer: ByteArray, offset: Int, length: Int, filePosition: Long, encrypt: Boolean,
    ) {
        if (length <= 0) return
        if (Looper.myLooper() == Looper.getMainLooper()) {
            throw IOException(
                "DartProxyCipherAdapter.transform called on the main thread; " +
                    "this would deadlock. Dart ciphers only work on background readers " +
                    "(playback / file encrypt); getMediaInfo may probe on the main thread.",
            )
        }

        // Frame: [dir: 0=decrypt, 1=encrypt][offset 8B BE][payload]. Direct buffer
        // required by send(). Do NOT flip(): the Android messenger dispatches
        // bytes [0, position), so flipping to position 0 sends an empty message.
        val request = ByteBuffer.allocateDirect(9 + length).order(ByteOrder.BIG_ENDIAN)
        request.put(if (encrypt) 1 else 0)
        request.putLong(filePosition)
        request.put(buffer, offset, length)

        val latch = CountDownLatch(1)
        // -2 = no/null reply, otherwise the reply's byte count. The reply buffer
        // is only guaranteed valid inside the callback, so copy there; writing
        // into the caller's array is safe because the caller is parked on the
        // latch until countDown() (happens-before via await).
        val repliedBytes = intArrayOf(-2)
        mainHandler.post {
            channel.send(request) { response ->
                if (response != null) {
                    repliedBytes[0] = response.remaining()
                    if (response.remaining() == length) {
                        response.get(buffer, offset, length)
                    }
                }
                latch.countDown()
            }
        }

        if (!latch.await(timeoutMs, TimeUnit.MILLISECONDS)) {
            throw IOException("Dart cipher timed out after ${timeoutMs}ms")
        }
        when (repliedBytes[0]) {
            length -> return
            -2 -> throw IOException("Dart cipher delegate returned an error or no data")
            else -> throw IOException(
                "Dart cipher returned ${repliedBytes[0]} bytes, expected $length",
            )
        }
    }
}
